tests/database/test_eth1_chaindb.py

```diff
@@ -84,13 +87,135 @@ def chain(chain_without_block_validation):
         return chain_without_block_validation
 
 
+def get_score(genesis_header, children):
+    return sum(header.difficulty for header in children) + genesis_header.difficulty
+
+
 def test_chaindb_add_block_number_to_hash_lookup(chaindb, block):
     block_number_to_hash_key = SchemaV1.make_block_number_to_hash_lookup_key(block.number)
     assert not chaindb.exists(block_number_to_hash_key)
+    assert chaindb.get_chain_gaps() == GENESIS_CHAIN_GAPS
     chaindb.persist_block(block)
     assert chaindb.exists(block_number_to_hash_key)
 
 
+@pytest.mark.parametrize(
+    'has_uncle, has_transaction, can_fetch_block, final_gaps',
+    (
+        # Uncle block gets de-canonicalized by a header that has another uncle
+        (True, False, False, (((4, 6),), 8),),
+        # Uncle block gets de-canonicalized by a header that has transactions
+        (False, True, False, (((4, 6),), 8),),
+        # Has uncle and transactions
+        (True, True, False, (((4, 6),), 8),),
+        # Uncle block gets de-canonicalized by a header that has no uncles nor transactions
+        # In this scenario, we do not need to re-open the gap because the new block is "trivial"
+        # which means it can be loaded by just the header alone.
+        (False, False, True, (((5, 6),), 8),),
+    )
+)
```

Comment:
@carver Since block gap tracking "inherits" from header gap tracking (There is no way to persist a block that bypasses the header gap tracking) my hope is that these cases are the only ones we have to cover....I hope :sweat_smile: